File: apps/web/src/utils/freeLimits.js
/**
 * Free-tier ceilings. FABTrades Pro removes all of them.
 *
 * These numbers are duplicated in `apps/mobile/lib/core/logic/free_limits.dart`
 * and pinned by `packages/contracts/free_limits.json`, which both test suites
 * read. Drift here is not cosmetic: mobile enforces the trade window by deleting
 * the oldest rows, so a higher cap on web would look to the customer like mobile
 * eating trades they had saved.
 *
 * Binder and want-list caps refuse new distinct cards (never trim existing rows).
 * Raising quantity on a card that is already listed is never capped — same as
 * mobile's `_canAddNewCard`.
 */
export const FreeLimits = {
    binderCards: 50,
    wantListCards: 50,
    savedTrades: 3,
    loanedCards: 1,
};

/**
 * Distinct-card cap for binder vs want list.
 *
 * @param {{ isWanted: boolean }} opts
 * @returns {number}
 */
export function cardsFor({ isWanted }) {
    return isWanted ? FreeLimits.wantListCards : FreeLimits.binderCards;
}

/**
 * Whether a free account may add another distinct card to binder or want list.
 *
 * Matches mobile: Pro always allowed; free accounts are blocked once they already
 * hold the cap. Callers that are only raising quantity on an existing entry should
 * not use this — that path is uncapped.
 *
 * @param {number} existingDistinctCount - How many distinct cards are already listed
 * @param {{ isWanted: boolean, isPro?: boolean }} opts
 * @returns {boolean}
 */
export function canAddDistinctCard(existingDistinctCount, { isWanted, isPro = false } = {}) {
    if (isPro) return true;
    return existingDistinctCount < cardsFor({ isWanted });
}

/**
 * How many of the oldest trades have to roll off to fit inside the free window.
 *
 * Saving a trade is never refused, on either client. Refusing would lose the
 * trade outright, and a history that quietly keeps the most recent is a much
 * gentler limit to run into than a save button that stops working.
 *
 * @param {number} total - Trades the account would hold after the save
 * @returns {number} Trades to tombstone, oldest first; 0 when inside the window
 */
export function tradesOverFreeLimit(total) {
    return Math.max(0, total - FreeLimits.savedTrades);
}

File: apps/mobile/lib/core/logic/free_limits.dart
/// Free-tier ceilings. FABTrades Pro removes all of them.
///
/// Numbers are pinned by `packages/contracts/free_limits.json` so web and mobile
/// cannot drift — especially the trade window, which both enforce by deleting
/// the oldest rows.
class FreeLimits {
  const FreeLimits._();

  /// Distinct cards a free binder holds. Raising the quantity of a card that's
  /// already there is never capped, so no existing entry can be stranded.
  static const binderCards = 5;

  /// Distinct cards a free want list holds.
  static const wantListCards = 25;

  /// Cards currently lent out across all lend groups.
  ///
  /// Counted by quantity, not distinct titles — lending two copies is two
  /// loaned cards. Borrowing is uncapped: the limit is on what you have out.
  static const loanedCards = 1;

  /// Trades a free account keeps.
  ///
  /// Confirming a trade is never blocked — the binder reconciliation it
  /// performs is core to the app, and refusing it would lose real information.
  /// History is a rolling window instead: the oldest entry drops off.
  static const savedTrades = 3;

  static int cardsFor({required bool isWanted}) =>
      isWanted ? wantListCards : binderCards;
}

/// How much of the free tier is currently in use, for upsell copy.
class FreeUsage {
  const FreeUsage({
    required this.binderCards,
    required this.wantListCards,
    required this.loanedCards,
    required this.savedTrades,
  });

  final int binderCards;
  final int wantListCards;
  final int loanedCards;
  final int savedTrades;

  /// The cap closest to being reached, as a 0–1 fraction. Drives whether the
  /// upgrade prompt mentions limits at all.
  double get pressure => [
        binderCards / FreeLimits.binderCards,
        wantListCards / FreeLimits.wantListCards,
        loanedCards / FreeLimits.loanedCards,
        savedTrades / FreeLimits.savedTrades,
      ].reduce((a, b) => a > b ? a : b);

  /// True once any cap is within a few entries, which is when nudging is
  /// useful rather than noise.
  bool get isNearAnyLimit => pressure >= 0.7;
}
